{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "postal-darwin",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from amino_acids import aa, codons, aa_table   # you may find these useful\n",
    "from load import load_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "quick-county",
   "metadata": {},
   "outputs": [],
   "source": [
    "def shuffle_string(s):\n",
    "    \"\"\"Shuffles the characters in the input string\n",
    "        NOTE: this is a helper function, you do not\n",
    "        have to modify this in any way \"\"\"\n",
    "    return ''.join(random.sample(s, len(s)))\n",
    "\n",
    "# YOU WILL START YOUR IMPLEMENTATION FROM HERE DOWN ###\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "secret-easter",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "tired-karen",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'T'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_complement(nucleotide):\n",
    "    \"\"\" Returns the complementary nucleotide\n",
    "\n",
    "        nucleotide: a nucleotide (A, C, G, or T) represented as a string\n",
    "        returns: the complementary nucleotide\n",
    "    >>> get_complement('A')\n",
    "    'T'\n",
    "    >>> get_complement('C')\n",
    "    'G'\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    \n",
    "    \n",
    "    if nucleotide == \"A\":\n",
    "        return \"T\"\n",
    "    elif nucleotide == \"C\":\n",
    "        return \"G\"\n",
    "    elif nucleotide == \"T\":\n",
    "        return \"A\"\n",
    "    elif nucleotide == \"G\":\n",
    "        return \"C\"\n",
    "    else:\n",
    "        return \"Not a nucleotide\"\n",
    "\n",
    "\n",
    "get_complement(\"A\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "specialized-constitutional",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TGAACGCGG'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_reverse_complement(dna):\n",
    "    \"\"\" Computes the reverse complementary sequence of DNA for the specfied DNA\n",
    "        sequence\n",
    "\n",
    "        dna: a DNA sequence represented as a string\n",
    "        returns: the reverse complementary DNA sequence represented as a string\n",
    "    >>> get_reverse_complement(\"ATGCCCGCTTT\")\n",
    "    'AAAGCGGGCAT'\n",
    "    >>> get_reverse_complement(\"CCGCGTTCA\")\n",
    "    'TGAACGCGG'\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "\n",
    "    complement = {'A': 'T', 'C': 'G', 'T': 'A', 'G': 'C'}\n",
    "\n",
    "    list1 = [\"\".join(complement[letter] for letter in dna)]\n",
    "\n",
    "    st1 = \"\"\n",
    "    for i in list1:\n",
    "        st1+=i\n",
    "\n",
    "    return st1[::-1]\n",
    "\n",
    "get_reverse_complement(\"CCGCGTTCA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "worth-jordan",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGAGA'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def rest_of_ORF(dna):\n",
    "    \"\"\" Takes a DNA sequence that is assumed to begin with a start\n",
    "        codon and returns the sequence up to but not including the\n",
    "        first in frame stop codon.  If there is no in frame stop codon,\n",
    "        returns the whole string.\n",
    "\n",
    "        dna: a DNA sequence\n",
    "        returns: the open reading frame represented as a string\n",
    "    >>> rest_of_ORF(\"ATGTGAA\")\n",
    "    'ATG'\n",
    "    >>> rest_of_ORF(\"ATGAGATAGG\")\n",
    "    'ATGAGA'\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    \n",
    "    # ORF1 CODE\n",
    "\n",
    "    Orf1 = []\n",
    "\n",
    "    for i in range(0, len(dna), 3 ):\n",
    "        Orf1.append( dna[0+int(i):3+int(i)] )\n",
    "\n",
    "#find and index all stops in dna\n",
    "    if  \"TGA\" in Orf1:\n",
    "        stop_at = -1\n",
    "        indexs_stop = []\n",
    "\n",
    "        while True:\n",
    "\n",
    "            try:\n",
    "                loc = Orf1.index(\"TGA\",stop_at+1)\n",
    "            except ValueError:\n",
    "                break\n",
    "            else:\n",
    "                indexs_stop.append(loc)\n",
    "                stop_at = loc\n",
    "\n",
    "        print(indexs_stop, \"\\n\")\n",
    "\n",
    "    if  \"TAA\" in Orf1:\n",
    "        stop_at = -1\n",
    "        indexs_stop = []\n",
    "\n",
    "        while True:\n",
    "\n",
    "            try:\n",
    "                loc = Orf1.index(\"TAA\",stop_at+1)\n",
    "            except ValueError:\n",
    "                break\n",
    "            else:\n",
    "                indexs_stop.append(loc)\n",
    "                stop_at = loc\n",
    "\n",
    "        print(indexs_stop, \"\\n\")\n",
    "\n",
    "    if  \"TAG\" in Orf1:\n",
    "        stop_at = -1\n",
    "        indexs_stop = []\n",
    "\n",
    "        while True:\n",
    "\n",
    "            try:\n",
    "                loc = Orf1.index(\"TAG\",stop_at+1)\n",
    "            except ValueError:\n",
    "                break\n",
    "            else:\n",
    "                indexs_stop.append(loc)\n",
    "                stop_at = loc        \n",
    "\n",
    "    if len(indexs_stop) > 0:\n",
    "        stop = indexs_stop[0]#stop index\n",
    "        return(\"\".join(Orf1[:stop]))\n",
    "    else:\n",
    "        return(\"\".join(Orf1[:]))\n",
    "    \n",
    "rest_of_ORF('ATGAGATAGG')      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "nominated-notion",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ATG', 'CAT', 'GAA', 'TGT', 'AGA', 'TAG', 'ATG', 'TGC', 'CC']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['ATGCATGAATGTAGA']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_all_ORFs_oneframe(dna):\n",
    "    \"\"\" Finds all non-nested open reading frames in the given DNA\n",
    "        sequence and returns them as a list.  This function should\n",
    "        only find ORFs that are in the default frame of the sequence\n",
    "        (i.e. they start on indices that are multiples of 3).\n",
    "        By non-nested we mean that if an ORF occurs entirely within\n",
    "        another ORF, it should not be included in the returned list of ORFs.\n",
    "\n",
    "        dna: a DNA sequence\n",
    "        returns: a list of non-nested ORFs\n",
    "    >>> find_all_ORFs_oneframe(\"ATGCATGAATGTAGATAGATGTGCCC\")\n",
    "    ['ATGCATGAATGTAG', 'ATGTGCCC']\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    dna = 'ATGCATGAATGTAGATAGATGTGCCC'\n",
    "    \n",
    "    Orf1 = []\n",
    "\n",
    "    for i in range(0, len(dna), 3 ):\n",
    "        Orf1.append( dna[0+int(i):3+int(i)] )\n",
    "    print(Orf1)\n",
    "\n",
    "    #find all start codons and index\n",
    "    start_at = -1\n",
    "    indexs_start = []\n",
    "    while True:\n",
    "        try:\n",
    "            loc = Orf1.index(\"ATG\",start_at+1)\n",
    "        except ValueError:\n",
    "            break\n",
    "        else:\n",
    "            indexs_start.append(loc)\n",
    "            start_at = loc\n",
    "\n",
    "    #find and index all stops in dna\n",
    "    if  \"TGA\" in Orf1:\n",
    "        stop_at = -1\n",
    "        indexs_stop = []\n",
    "\n",
    "        while True:\n",
    "\n",
    "            try:\n",
    "                loc = Orf1.index(\"TGA\",stop_at+1)\n",
    "            except ValueError:\n",
    "                break\n",
    "            else:\n",
    "                indexs_stop.append(loc)\n",
    "                stop_at = loc\n",
    "\n",
    "    if  \"TAA\" in Orf1:\n",
    "        stop_at = -1\n",
    "        indexs_stop = []\n",
    "\n",
    "        while True:\n",
    "\n",
    "            try:\n",
    "                loc = Orf1.index(\"TAA\",stop_at+1)\n",
    "            except ValueError:\n",
    "                break\n",
    "            else:\n",
    "                indexs_stop.append(loc)\n",
    "                stop_at = loc\n",
    "\n",
    "    if  \"TAG\" in Orf1:\n",
    "        stop_at = -1\n",
    "        indexs_stop = []\n",
    "\n",
    "        while True:\n",
    "\n",
    "            try:\n",
    "                loc = Orf1.index(\"TAG\",stop_at+1)\n",
    "            except ValueError:\n",
    "                break\n",
    "            else:\n",
    "                indexs_stop.append(loc)\n",
    "                stop_at = loc        \n",
    "    \n",
    "    for i, j in zip(indexs_start, indexs_stop):\n",
    "        return ([\"\".join(Orf1[i:j])])\n",
    "\n",
    "find_all_ORFs_oneframe('ATGCATGAATGTAGATAGATGTGCCC')   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "spectacular-whole",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ORF1 IS ----> ['ATG', 'CAT', 'GAA', 'TGT', 'AG'] \n",
      "\n",
      "ORF2 IS ----> ['TGC', 'ATG', 'AAT', 'GTA', 'G'] \n",
      "\n",
      "ORF3 IS ----> ['GCA', 'TGA', 'ATG', 'TAG', ''] \n",
      "\n",
      "['ATGCATGAATGTAG']\n",
      "['ATGAATGTAG']\n",
      "['ATGTAG']\n"
     ]
    }
   ],
   "source": [
    "def find_all_ORFs(dna):\n",
    "    \"\"\" Finds all non-nested open reading frames in the given DNA sequence in\n",
    "        all 3 possible frames and returns them as a list.  By non-nested we\n",
    "        mean that if an ORF occurs entirely within another ORF and they are\n",
    "        both in the same frame, it should not be included in the returned list\n",
    "        of ORFs.\n",
    "\n",
    "        dna: a DNA sequence\n",
    "        returns: a list of non-nested ORFs\n",
    "\n",
    "    >>> find_all_ORFs(\"ATGCATGAATGTAG\")\n",
    "    ['ATGCATGAATGTAG', 'ATGAATGTAG', 'ATG']\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    \n",
    "    # ORF1 \n",
    "    Orf1 = []\n",
    "\n",
    "    for i in range(0, len(dna), 3 ):\n",
    "        Orf1.append( dna[0+int(i):3+int(i)] )\n",
    "    print(\"ORF1 IS ---->\", Orf1, \"\\n\")\n",
    "\n",
    "    #ORF2\n",
    "\n",
    "    Orf2 = []\n",
    "\n",
    "    for i in range(0, len(dna), 3 ):\n",
    "        Orf2.append( dna[1+int(i):4+int(i)] )\n",
    "    print(\"ORF2 IS ---->\", Orf2, \"\\n\")\n",
    "\n",
    "    #ORF3\n",
    "\n",
    "    Orf3 = []\n",
    "\n",
    "    for i in range(0, len(dna), 3 ):\n",
    "        Orf3.append( dna[2+int(i):5+int(i)] )\n",
    "    print(\"ORF3 IS ---->\", Orf3, \"\\n\")\n",
    "\n",
    "\n",
    "    all_orfs = [Orf1, Orf2, Orf3]\n",
    "\n",
    "    for rf in all_orfs:\n",
    "\n",
    "        start_at = -1\n",
    "        indexs_start = []\n",
    "\n",
    "        while True:\n",
    "\n",
    "            try:\n",
    "                loc = rf.index(\"ATG\",start_at+1)\n",
    "            except ValueError:\n",
    "                break\n",
    "            else:\n",
    "                indexs_start.append(loc)\n",
    "                start_at = loc\n",
    "\n",
    "        nest_list = []\n",
    "        for i in indexs_start:\n",
    "            nest_list.append(\"\".join(rf[i:]))\n",
    "        print(nest_list)\n",
    "        \n",
    "    return\n",
    "\n",
    "find_all_ORFs('ATGCATGAATGTAG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "distributed-coast",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ORF1 IS ----> ['ATG', 'CGA', 'ATG', 'TAG', 'CAT', 'CAA', 'A'] \n",
      "\n",
      "ORF2 IS ----> ['TGC', 'GAA', 'TGT', 'AGC', 'ATC', 'AAA', ''] \n",
      "\n",
      "ORF3 IS ----> ['GCG', 'AAT', 'GTA', 'GCA', 'TCA', 'AA', ''] \n",
      "\n",
      "ORF4 IS ----> ['TTT', 'GAT', 'GCT', 'ACA', 'TTC', 'GCA', 'T'] \n",
      "\n",
      "ORF5 IS ----> ['TTG', 'ATG', 'CTA', 'CAT', 'TCG', 'CAT', ''] \n",
      "\n",
      "ORF6 IS ----> ['TGA', 'TGC', 'TAC', 'ATT', 'CGC', 'AT', ''] \n",
      "\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'j' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-9-1d771d7d2c89>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     83\u001b[0m     \u001b[1;32mreturn\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     84\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 85\u001b[1;33m \u001b[0mfind_all_ORFs_both_strands\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'ATGCGAATGTAGCATCAAA'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-9-1d771d7d2c89>\u001b[0m in \u001b[0;36mfind_all_ORFs_both_strands\u001b[1;34m(dna)\u001b[0m\n\u001b[0;32m     78\u001b[0m         \u001b[0mnest_list\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     79\u001b[0m         \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mindexs_start\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 80\u001b[1;33m             \u001b[0mnest_list\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"\"\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mrf\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m:\u001b[0m\u001b[0mj\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     81\u001b[0m         \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnest_list\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     82\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'j' is not defined"
     ]
    }
   ],
   "source": [
    "def find_all_ORFs_both_strands(dna):\n",
    "    \"\"\" Finds all non-nested open reading frames in the given DNA sequence on both\n",
    "        strands.\n",
    "\n",
    "        dna: a DNA sequence\n",
    "        returns: a list of non-nested ORFs\n",
    "    >>> find_all_ORFs_both_strands(\"ATGCGAATGTAGCATCAAA\")\n",
    "    ['ATGCGAATG', 'ATGCTACATTCGCAT']\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    \n",
    "    # ORF1 \n",
    "    Orf1 = []\n",
    "\n",
    "    for i in range(0, len(dna), 3 ):\n",
    "        Orf1.append( dna[0+int(i):3+int(i)] )\n",
    "    print(\"ORF1 IS ---->\", Orf1, \"\\n\")\n",
    "\n",
    "    #ORF2\n",
    "\n",
    "    Orf2 = []\n",
    "\n",
    "    for i in range(0, len(dna), 3 ):\n",
    "        Orf2.append( dna[1+int(i):4+int(i)] )\n",
    "    print(\"ORF2 IS ---->\", Orf2, \"\\n\")\n",
    "\n",
    "    #ORF3\n",
    "\n",
    "    Orf3 = []\n",
    "\n",
    "    for i in range(0, len(dna), 3 ):\n",
    "        Orf3.append( dna[2+int(i):5+int(i)] )\n",
    "    print(\"ORF3 IS ---->\", Orf3, \"\\n\")\n",
    "\n",
    "    \n",
    "    rev_dna = get_reverse_complement(dna)\n",
    "    \n",
    "    #ORF4\n",
    "    Orf4 = []\n",
    "\n",
    "    for i in range(0, len(rev_dna), 3 ):\n",
    "        Orf4.append( rev_dna[0+int(i):3+int(i)] )\n",
    "    print(\"ORF4 IS ---->\", Orf4, \"\\n\")\n",
    "\n",
    "    #ORF5\n",
    "\n",
    "    Orf5 = []\n",
    "\n",
    "    for i in range(0, len(rev_dna), 3 ):\n",
    "        Orf5.append( rev_dna[1+int(i):4+int(i)] )\n",
    "    print(\"ORF5 IS ---->\", Orf5, \"\\n\")\n",
    "\n",
    "    #ORF6\n",
    "\n",
    "    Orf6 = []\n",
    "\n",
    "    for i in range(0, len(rev_dna), 3 ):\n",
    "        Orf6.append( rev_dna[2+int(i):5+int(i)] )\n",
    "    print(\"ORF6 IS ---->\", Orf6, \"\\n\")\n",
    "    \n",
    "    all_orfs = [Orf1, Orf2, Orf3, Orf4, Orf5, Orf6]\n",
    "    \n",
    "    for rf in all_orfs:\n",
    "\n",
    "        start_at = -1\n",
    "        indexs_start = []\n",
    "\n",
    "        while True:\n",
    "\n",
    "            try:\n",
    "                loc = rf.index(\"ATG\",start_at+1)\n",
    "            except ValueError:\n",
    "                break\n",
    "            else:\n",
    "                indexs_start.append(loc)\n",
    "                start_at = loc\n",
    "\n",
    "        nest_list = []\n",
    "        for i in indexs_start:\n",
    "            nest_list.append(\"\".join(rf[i:j]))\n",
    "        print(nest_list)\n",
    "        \n",
    "    return\n",
    "\n",
    "find_all_ORFs_both_strands('ATGCGAATGTAGCATCAAA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "voluntary-phrase",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ATGCTACATTCGCAT'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def longest_ORF(dna):\n",
    "    \"\"\" Finds the longest ORF on both strands of the specified DNA and returns it\n",
    "        as a string\n",
    "    >>> longest_ORF(\"ATGCGAATGTAGCATCAAA\")\n",
    "    'ATGCTACATTCGCAT'\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "\n",
    "    import re\n",
    "\n",
    "\n",
    "    if \"TAG\" in dna or \"TAA\" in dna or \"TGA\" in dna:\n",
    "        f = (max(re.findall(r'ATG(?:(?!TAA|TAG|TGA)...)*(?:TAA|TAG|TGA)',dna), key = len))\n",
    "    else:\n",
    "        f = (max(re.findall(r'ATG(?:...)*(?:.*)',dna), key = len))\n",
    "\n",
    "    revers = get_reverse_complement(dna)\n",
    "\n",
    "    if \"TAG\" in revers or \"TAA\" in revers or \"TGA\" in revers:\n",
    "        r = (max(re.findall(r'ATG(?:...)*(?:.*)',revers), key = len))\n",
    "\n",
    "    else:\n",
    "        r= (max(re.findall(r'ATG(?:(?!TAA|TAG|TGA)...)*(?:TAA|TAG|TGA)',revers), key = len))\n",
    "\n",
    "\n",
    "    if len(f) > len(r):\n",
    "        return (f)\n",
    "\n",
    "    else:\n",
    "        return (r)\n",
    "    \n",
    "longest_ORF('ATGCGAATGTAGCATCAAA')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "occupied-portuguese",
   "metadata": {},
   "outputs": [],
   "source": [
    "def longest_ORF_noncoding(dna, num_trials):\n",
    "    \"\"\" Computes the maximum length of the longest ORF over num_trials shuffles\n",
    "        of the specfied DNA sequence\n",
    "\n",
    "        dna: a DNA sequence\n",
    "        num_trials: the number of random shuffles\n",
    "        returns: the maximum length longest ORF \"\"\"\n",
    "    # TODO: implement this\n",
    "\n",
    "    import random\n",
    "    import re\n",
    "\n",
    "    longest_orf_shuffled =[]\n",
    "    for iteration in range(1000):\n",
    "    #     print(iteration)\n",
    "        s=\"ATGCGAATGTAGCATCAAACAAGAAGGATTGACAATCT\"\n",
    "        dna = ''.join(random.sample(s,len(s)))\n",
    "\n",
    "\n",
    "        if \"ATG\" in dna:\n",
    "            fowards = (max(re.findall(r'ATG(?:...)*(?:.*)',dna), key = len))\n",
    "\n",
    "\n",
    "\n",
    "        revers = get_reverse_complement(dna)\n",
    "\n",
    "        if \"ATG\" in revers:\n",
    "            rev = (max(re.findall(r'ATG(?:...)*(?:.*)',revers), key = len))\n",
    "\n",
    "\n",
    "        if len(fowards) > len(rev):\n",
    "            longest_orf_shuffled.append(foward)\n",
    "        else:\n",
    "            longest_orf_shuffled.append(rev)\n",
    "\n",
    "    max(longest_orf_shuffled, key=len)\n",
    "        \n",
    "    return longest_ORF_noncoding\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "brown-debate",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ORF IS ----> ['ATG', 'CCC', 'GCT', 'TT'] \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'M'"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def coding_strand_to_AA(dna):\n",
    "    \"\"\" Computes the Protein encoded by a sequence of DNA.  This function\n",
    "        does not check for start and stop codons (it assumes that the input\n",
    "        DNA sequence represents an protein coding region).\n",
    "\n",
    "        dna: a DNA sequence represented as a string\n",
    "        returns: a string containing the sequence of amino acids encoded by the\n",
    "                 the input DNA fragment\n",
    "\n",
    "        >>> coding_strand_to_AA(\"ATGCGA\")\n",
    "        'MR'\n",
    "        >>> coding_strand_to_AA(\"ATGCCCGCTTT\")\n",
    "        'MPA'\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    \n",
    "    aa = ['F', 'L', 'I', 'M', 'V', 'S', 'P', 'T', 'A', 'Y',\n",
    "          '|', 'H', 'Q', 'N', 'K', 'D', 'E', 'C', 'W', 'R',\n",
    "          'G']\n",
    "\n",
    "    codons = [['TTT', 'TTC'],\n",
    "              ['TTA', 'TTG', 'CTT', 'CTC', 'CTA', 'CTG'],\n",
    "              ['ATT', 'ATC', 'ATA'],\n",
    "              ['ATG'],\n",
    "              ['GTT', 'GTC', 'GTA', 'GTG'],\n",
    "              ['TCT', 'TCC', 'TCA', 'TCG', 'AGT', 'AGC'],\n",
    "              ['CCT', 'CCC', 'CCA', 'CCG'],\n",
    "              ['ACT', 'ACC', 'ACA', 'ACG'],\n",
    "              ['GCT', 'GCC', 'GCA', 'GCG'],\n",
    "              ['TAT', 'TAC'],\n",
    "              ['TAA', 'TAG', 'TGA'],\n",
    "              ['CAT', 'CAC'],\n",
    "              ['CAA', 'CAG'],\n",
    "              ['AAT', 'AAC'],\n",
    "              ['AAA', 'AAG'],\n",
    "              ['GAT', 'GAC'],\n",
    "              ['GAA', 'GAG'],\n",
    "              ['TGT', 'TGC'],\n",
    "              ['TGG'],\n",
    "              ['CGT', 'CGC', 'CGA', 'CGG', 'AGA', 'AGG'],\n",
    "              ['GGT', 'GGC', 'GGA', 'GGG']]\n",
    "\n",
    "    # create a dictionary lookup table for mapping codons into amino acids\n",
    "    aa_table = {}\n",
    "    for i in range(len(aa)):\n",
    "        for codon in codons[i]:\n",
    "            aa_table[codon] = aa[i]\n",
    "\n",
    "    #     print (aa_table)\n",
    "\n",
    "    Orf = []\n",
    "\n",
    "    for i in range(0, len(dna), 3 ):\n",
    "        Orf.append( dna[0+int(i):3+int(i)] )\n",
    "    print(\"ORF IS ---->\", Orf, \"\\n\")\n",
    "\n",
    "    count = 0\n",
    "    amino_seq = (''.join([aa_table[dna[pos:pos + 3]] for pos in range (count, len(Orf) -2, 3)]))\n",
    "    return (amino_seq)\n",
    "\n",
    "    \n",
    "coding_strand_to_AA('ATGCCCGCTTT')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "american-nepal",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ORF1 IS ----> ['ATG', 'AAG', 'CAT', 'TAA', 'ACC', 'ACA', 'AAG', 'ATG', 'TGT', 'AGA', 'TGA', 'GCA', 'CT'] \n",
      "\n",
      "ORF2 IS ----> ['TGA', 'AGC', 'ATT', 'AAA', 'CCA', 'CAA', 'AGA', 'TGT', 'GTA', 'GAT', 'GAG', 'CAC', 'T'] \n",
      "\n",
      "ORF3 IS ----> ['GAA', 'GCA', 'TTA', 'AAC', 'CAC', 'AAA', 'GAT', 'GTG', 'TAG', 'ATG', 'AGC', 'ACT', ''] \n",
      "\n",
      "ORF4 IS ----> ['AGT', 'GCT', 'CAT', 'CTA', 'CAC', 'ATC', 'TTT', 'GTG', 'GTT', 'TAA', 'TGC', 'TTC', 'AT'] \n",
      "\n",
      "ORF5 IS ----> ['GTG', 'CTC', 'ATC', 'TAC', 'ACA', 'TCT', 'TTG', 'TGG', 'TTT', 'AAT', 'GCT', 'TCA', 'T'] \n",
      "\n",
      "ORF6 IS ----> ['TGC', 'TCA', 'TCT', 'ACA', 'CAT', 'CTT', 'TGT', 'GGT', 'TTA', 'ATG', 'CTT', 'CAT', ''] \n",
      "\n",
      "['ATGAAGCATTAAACCACAAAGATGTGTAGATGAGCACT', 'ATGTGTAGATGAGCACT']\n",
      "[]\n",
      "['ATGAGCACT']\n",
      "[]\n",
      "[]\n",
      "['ATGCTTCAT']\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "'NoneType' object is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-107-f032ae915a20>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     13\u001b[0m            \u001b[1;32mreturn\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mall_genes\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcoding_strand_to_AA\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     14\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 15\u001b[1;33m \u001b[0mgene_finder\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'ATGCCCGCTTT'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-107-f032ae915a20>\u001b[0m in \u001b[0;36mgene_finder\u001b[1;34m(dna)\u001b[0m\n\u001b[0;32m     10\u001b[0m     \u001b[0mall_genes\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 12\u001b[1;33m     \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mfind_all_ORFs_both_strands\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdna\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     13\u001b[0m            \u001b[1;32mreturn\u001b[0m \u001b[1;33m(\u001b[0m\u001b[0mall_genes\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcoding_strand_to_AA\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     14\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: 'NoneType' object is not iterable"
     ]
    }
   ],
   "source": [
    "def gene_finder(dna):\n",
    "    \"\"\" Returns the amino acid sequences that are likely coded by the specified dna\n",
    "\n",
    "        dna: a DNA sequence\n",
    "        returns: a list of all amino acid sequences coded by the sequence dna.\n",
    "    \"\"\"\n",
    "    # TODO: implement this\n",
    "    dna = 'ATGAAGCATTAAACCACAAAGATGTGTAGATGAGCACT'\n",
    "\n",
    "    all_genes = []\n",
    "\n",
    "    for i in find_all_ORFs_both_strands(dna):\n",
    "           return (all_genes.append(coding_strand_to_AA(i)))\n",
    "        \n",
    "gene_finder('ATGCCCGCTTT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "stuck-moses",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-109-cba97e046d59>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-109-cba97e046d59>\"\u001b[1;36m, line \u001b[1;32m1\u001b[0m\n\u001b[1;33m    def menu:\u001b[0m\n\u001b[1;37m            ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "def menu:\n",
    "    \n",
    "    \"\"\"\n",
    "    Select an option\n",
    "  [R] Display read counts in each Chromosome\n",
    "  [D] Display the average read quality for each chromosome\n",
    "  [X] Exit\n",
    "  \n",
    "    \"\"\"\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "organizational-metadata",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " WELCOME TO GENE FINDER\n",
      "\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "\n",
      "                        Select an option\n",
      "                      [A] Get complement\n",
      "                      [B] Get Reverse Complement\n",
      "                      [C] Get Rest of ORF\n",
      "                      [D] Find All ORFS in One Frame\n",
      "                      [E] Find All ORFS\n",
      "                      [F] Find All ORFS Both Strands\n",
      "                      [G] Find Longest ORF\n",
      "                      [H] Find Longest ORF Noncoding\n",
      "                      [I] Convert Strand to AA\n",
      "                      [J] Gene Finder\n",
      "                      [Q] Exit\n",
      "                      \n",
      "                      Please enter your choice:  a\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'nucleotide' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-3-23c78edaac2f>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     53\u001b[0m         \u001b[0mmenu\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     54\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 55\u001b[1;33m \u001b[0mmain\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;32m<ipython-input-3-23c78edaac2f>\u001b[0m in \u001b[0;36mmain\u001b[1;34m()\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mmain\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m     \u001b[0mmenu\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      6\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-3-23c78edaac2f>\u001b[0m in \u001b[0;36mmenu\u001b[1;34m()\u001b[0m\n\u001b[0;32m     27\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     28\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0mchoice\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m\"A\"\u001b[0m \u001b[1;32mor\u001b[0m \u001b[0mchoice\u001b[0m \u001b[1;33m==\u001b[0m\u001b[1;34m\"a\"\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 29\u001b[1;33m         \u001b[0mget_complement\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnucleotide\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     30\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0mchoice\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m\"B\"\u001b[0m \u001b[1;32mor\u001b[0m \u001b[0mchoice\u001b[0m \u001b[1;33m==\u001b[0m\u001b[1;34m\"b\"\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     31\u001b[0m         \u001b[0mget_reverse_complement\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdna\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m: name 'nucleotide' is not defined"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "import sys\n",
    "\n",
    "def main():\n",
    "    menu()\n",
    "\n",
    "\n",
    "def menu():\n",
    "    print(\" WELCOME TO GENE FINDER\")        \n",
    "    print()\n",
    "\n",
    "    choice = input(\"\"\"\n",
    "                        Select an option\n",
    "                      [A] Get complement\n",
    "                      [B] Get Reverse Complement\n",
    "                      [C] Get Rest of ORF\n",
    "                      [D] Find All ORFS in One Frame\n",
    "                      [E] Find All ORFS\n",
    "                      [F] Find All ORFS Both Strands\n",
    "                      [G] Find Longest ORF\n",
    "                      [H] Find Longest ORF Noncoding\n",
    "                      [I] Convert Strand to AA\n",
    "                      [J] Gene Finder\n",
    "                      [Q] Exit\n",
    "                      \n",
    "                      Please enter your choice: \"\"\")\n",
    "\n",
    "    if choice == \"A\" or choice ==\"a\":\n",
    "        get_complement(nucleotide)\n",
    "    elif choice == \"B\" or choice ==\"b\":\n",
    "        get_reverse_complement(dna)\n",
    "    elif choice==\"C\" or choice==\"c\":\n",
    "        rest_of_ORF(dna)\n",
    "    elif choice == \"D\" or choice ==\"d\":\n",
    "        find_all_ORFs_oneframe(dna)\n",
    "    elif choice==\"E\" or choice==\"e\":\n",
    "        find_all_ORFs(dna)\n",
    "    elif choice == \"F\" or choice ==\"\":\n",
    "        find_all_ORFs_both_strands(dna)\n",
    "    elif choice==\"G\" or choice==\"g\":\n",
    "        longest_ORF(dna)\n",
    "    elif choice == \"H\" or choice ==\"h\":\n",
    "        longest_ORF_noncoding(dna, num_trials)\n",
    "    elif choice==\"I\" or choice==\"i\":\n",
    "        coding_strand_to_AA(dna)  \n",
    "    elif choice==\"J\" or choice==\"j\":\n",
    "        gene_finder(dna)\n",
    "    elif choice==\"Q\" or choice==\"q\":\n",
    "        sys.exit\n",
    "    else:\n",
    "        print(\"You must only select either A,B,C,D,E,F,G,H,I,J,Q \")\n",
    "        print(\"Please try again\")\n",
    "        menu()\n",
    "\n",
    "main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
